draft/celery-5-high-level-architecture.rst

```diff
@@ -0,0 +1,1900 @@
+which either listen or poll for incoming data from a data source such as a
+database, a file system or an HTTP(S) endpoint.
+
+These services produce :term:`Document Messages <Document Message>`.
+
+Tasks which are subscribed to Data Sources will receive the raw document
+messages for further processing.
+
+.. admonition:: Example
+
+  We'd like to design a feature which locks Github issues immediately after
+  they are closed.
+
+  Github uses Webhooks to notify us when an issue is closed.
+
+  We set up a Data Source which starts an HTTPS server and expects incoming
+  HTTP requests on an endpoint.
+
+  Whenever a request arrives a :term:`Document Message` is published.
+
+Data Sinks
+~~~~~~~~~~
+
+A result from a :ref:`task <draft/celery-5-high-level-architecture:Tasks>`
+produces a :term:`Document Message` which a Data Sink or multiple Data Sinks
+consume.
+
+These :term:`Document Messages <Document Message>` are then stored in the Sinks
+the task is registered to.
+
+.. admonition:: Example
+
+  We have a task which calculates the hourly average impressions of a user's
+  post over a period of time.
+
+  The BI team requires the data to be inserted to `BigQuery <https://cloud.google.com/bigquery/>`_
+  because it uses it to research the effectiveness of users posts.
+
+  However, the user-facing post analytics dashboard also requires this data
+  and the team that maintains it doesn't want to use BigQuery because it is not
+  a cost-effective solution and because they already use `MongoDB <https://mongodb.com/>`_
+  to store all user-facing analytics data.
+
+  To resolve the issue we declare that the task routes it's results to two data
+  sinks. One for the BI team and the other for the analytics team.
+
+  Each data sink is configured to insert the data to a specific table
+  or collection.
+
+Controller
+----------
+
+The Controller is responsible for managing the lifecycle of all other Celery
+components.
+
+Celery 5 is a more complex system with multiple components and will often be
+deployed in high throughput, highly available production systems.
+
+The introduction of multiple components require us to have another component
+that manages the entire Celery cluster.
+
+During the lifecycle of a worker the Controller also manages and optimizes the
+execution of tasks to ensure we maximize the utilization of all our resources
+and to prevent expected errors.
+
+.. note::
+
+  The Controller is meant to be run as a user service.
+  If the Controller is run with root privileges, a log message with
+  the warning level will be emitted.
+
+Foreman
++++++++
+
+The Foreman service is responsible for spawning the :ref:`Workers <draft/celery-5-high-level-architecture:Worker>`,
+:ref:`Routers <draft/celery-5-high-level-architecture:Router>` and
+:ref:`Schedulers <draft/celery-5-high-level-architecture:Scheduler>`.
+
+By default, the Foreman service creates sub-processes for
```

Comment:
So these will now be entirely distinct processes?

Reply:
Yes.
The Controller will simply start them.
There will be no communication between them at all.